src/components/interface/cbufp/cbuf_c.c

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright 2010 by The George Washington University.  All rights reserved.
+ *
+ * Redistribution of this file is permitted under the GNU General
+ * Public License v2.
+ *
+ * Author: Gabriel Parmer, gparmer@gwu.edu, 2010
+ * Updated by Qi Wang and Jiguo Song, 2011
+ * Updated and simplified by removing sub-page allocations, Gabe Parmer, 2012
+ */
+
+#include <cos_component.h>
+#include <print.h>
+#include <cbuf.h>
+#include <cbufp.h>
+#include <cbuf_vect.h>
+#include <cos_vect.h>
+#include <cos_debug.h>
+#include <cos_alloc.h>
+#include <valloc.h>
+
+#define CSLAB_ALLOC(sz)   alloc_page()
+#define CSLAB_FREE(x, sz) free_page(x)
+#include <cslab.h>
+
+/* 
+ * All of the structures that must be compiled (only once) into each
+ * component that uses cbufs.
+ *
+ * I'd like to avoid having to make components initialize this data,
+ * so try and use structures that by default are initialized as all
+ * zero. 
+ */
+extern struct cos_component_information cos_comp_info;
+
+CVECT_CREATE_STATIC(meta_cbuf);
+PERCPU_VAR(cbuf_alloc_freelists);
+
+/*** Slow paths for each cbuf operation ***/
+
+/* 
+ * Precondition: cbuf lock is taken.
+ *
+ * A component has tried to map a cbuf_t to a buffer, but that cbuf
+ * isn't mapping into the component.  The component's cache of cbufs
+ * had a miss.
+ */
+int 
+__cbuf_2buf_miss(int cbid, int len)
+{
+	struct cbuf_meta *mc;
+	int ret;
+	void *h;
+
+	/* 
+	 * FIXME: This can lead to a DOS where the client passes all
+	 * possible cbids to be cbuf2bufed, which will allocate the
+	 * entire 4M of meta vectors in this component.  Yuck.
+	 * 
+	 * Solution: the security mechanisms of the kernel (preventing
+	 * cbids being passed to this component if they don't already
+	 * belong to the client) should fix this.
+	 */
+	mc = cbuf_vect_lookup_addr(cbid_to_meta_idx(cbid));
+	/* ...have to expand the cbuf_vect */
+	if (unlikely(!mc)) {
+		if (cbuf_vect_expand(&meta_cbuf, cbid_to_meta_idx(cbid))) BUG();    //???
+		mc = cbuf_vect_lookup_addr(cbid_to_meta_idx(cbid));
+		assert(mc);
+	}
+	ret = cbufp_retrieve(cos_spd_id(), cbid, len);
+	if (unlikely(ret < 0 || mc->sz < (len >> PAGE_ORDER))) return -1;
+	assert(CBUFM_GET_PTR(mc));
+
+	return 0;
+}
+
+static inline int
+__cbufp_alloc_slow(int cbid, int size, int *len, int *error)
+{
+	int amnt = 0, i;
+	static struct cbufp_shared_page *csp = NULL;
+
+	assert(cbid <= 0);
+	if (cbid == 0) {
+		struct cbuf_meta *cm;
+		struct cbufp_ring_element el;
+		if (!csp) csp = (struct cbufp_shared_page*)cbufp_map_collect(cos_spd_id());
+		/* Do a garbage collection */
+		amnt = cbufp_collect(cos_spd_id(), size);
```

Comment:
An open question:  Do we want to collect only the cbufs of the specific size, or do we want to collect all cbufs, with a priority for those of the given size?  This is a trade-off between amortizing the cbuf_collect calls, and doing computation that is guaranteed to not be useful at the current time.